File: bin/my_scripts/utils/format_merge_lines.swift
// Swift, known for its use in iOS and macOS application development, is also
// capable of scripting and handling file operations and string manipulation.

import Foundation

func checkFirstLine(_ line: String) -> Bool {
    let tempLine = line.trimmingCharacters(in: .whitespaces)
    return (tempLine.isEmpty || tempLine.hasPrefix("/") || tempLine.hasPrefix("*")) && !tempLine.hasPrefix("#")
}

func checkForComment(_ line: String) -> Bool {
    let trimmedLine = line.trimmingCharacters(in: .whitespaces)
    return trimmedLine.hasPrefix("/") || trimmedLine.hasPrefix("*")
}

func checkForContent(_ line: String) -> Bool {
    let trimmedLine = line.trimmingCharacters(in: .whitespaces)
    return !trimmedLine.isEmpty && !checkForComment(trimmedLine) && !trimmedLine.hasPrefix("#")
}

func checkForParen(_ line: String) -> Bool {
    return line.contains("(")
}

func checkForEndParen(_ line: String) -> Bool {
    return line.trimmingCharacters(in: .whitespaces).hasSuffix(")")
}

func checkForFuncStart(_ line: String) -> Bool {
    return line.replacingOccurrences(of: "\n", with: "").hasPrefix("{")
}

func processFile(_ filename: String) -> Int {
    do {
        let fileContents = try String(contentsOfFile: filename, encoding: .utf8)
        let lines = fileContents.components(separatedBy: "\n")
        var newLines = [String]()
        var changeCounter = 0
        var i = 0

        while i < lines.count {
            if i < lines.count - 3 && checkFirstLine(lines[i]) && checkForContent(lines[i + 1]) && checkForEndParen(lines[i + 2]) && checkForFuncStart(lines[i + 3]) {
                newLines.append(lines[i])
                newLines.append("\(lines[i + 1].trimmingCharacters(in: .whitespaces)) \(lines[i + 2].trimmingCharacters(in: .whitespaces))")
                newLines.append(lines[i + 3])
                i += 4
                changeCounter += 1
            } else {
                newLines.append(lines[i])
                i += 1
            }
        }

        try newLines.joined(separator: "\n").write(toFile: filename, atomically: true, encoding: .utf8)
        return changeCounter
    } catch {
        print("An error occurred: \(error)")
        return 0
    }
}

let fileManager = FileManager.default
let dirPath = "./" // Current directory

do {
    let items = try fileManager.contentsOfDirectory(atPath: dirPath)
    var totalChangeCount = 0

    for item in items {
        if item.hasSuffix(".c") {
            print("Processing file: \(item)")
            let fileChangeCount = processFile("\(dirPath)/\(item)")
            print("Done! Changes done: \(fileChangeCount)")
            totalChangeCount += fileChangeCount
        }
    }

    print("Done processing files! Total changes done: \(totalChangeCount)")
} catch {
    print("Failed to read contents of the directory: \(error)")
}

File: bin/my_scripts/utils/format_merge_lines.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "path/filepath"
    "strings"
)

func checkFirstLine(line string) bool {
    tempLine := strings.TrimSpace(line)
    return (tempLine == "" || strings.HasPrefix(tempLine, "/") || strings.HasPrefix(tempLine, "*")) && !strings.HasPrefix(tempLine, "#")
}

func checkForComment(line string) bool {
    return strings.HasPrefix(strings.TrimSpace(line), "/") || strings.HasPrefix(strings.TrimSpace(line), "*")
}

func checkForContent(line string) bool {
    return strings.TrimSpace(line) != "" && !checkForComment(line) && !strings.HasPrefix(strings.TrimSpace(line), "#")
}

func checkForParen(line string) bool {
    return strings.Contains(strings.TrimSpace(line), "(")
}

func checkForEndParen(line string) bool {
    return strings.HasSuffix(strings.TrimSpace(line), ")")
}

func checkForFuncStart(line string) bool {
    return strings.HasPrefix(strings.Replace(line, "\n", "", -1), "{")
}

func processFile(filename string) int {
    file, err := os.Open(filename)
    if err != nil {
        fmt.Println("Error opening file:", err)
        return 0
    }
    defer file.Close()

    scanner := bufio.NewScanner(file)
    var lines []string
    for scanner.Scan() {
        lines = append(lines, scanner.Text())
    }

    var newLines []string
    changeCounter := 0
    i := 0
    for i < len(lines) {
        if i < len(lines)-3 && checkFirstLine(lines[i]) && checkForContent(lines[i+1]) && checkForEndParen(lines[i+2]) && checkForFuncStart(lines[i+3]) {
            newLines = append(newLines, lines[i])
            newLines = append(newLines, lines[i+1] + " " + lines[i+2])
            newLines = append(newLines, lines[i+3])
            i += 4
            changeCounter++
        } else if i < len(lines)-4 && checkFirstLine(lines[i]) && checkForContent(lines[i+1]) && checkForParen(lines[i+2]) && checkForEndParen(lines[i+3]) && checkForFuncStart(lines[i+4]) {
            newLines = append(newLines, lines[i])
            newLines = append(newLines, lines[i+1] + " " + lines[i+2])
            newLines = append(newLines, lines[i+3])
            newLines = append(newLines, lines[i+4])
            i += 5
            changeCounter++
        } else if i < len(lines)-5 && checkFirstLine(lines[i]) && checkForContent(lines[i+1]) && checkForParen(lines[i+2]) && checkForContent(lines[i+3]) && checkForEndParen(lines[i+4]) && checkForFuncStart(lines[i+5]) {
            fmt.Printf("Found function with three lines of arguments, see: %s\n", lines[i+1]+lines[i+2]+lines[i+3]+lines[i+4])
            newLines = append(newLines, lines[i])
            i++
        } else {
            newLines = append(newLines, lines[i])
            i++
        }
    }

    // Write the processed lines back to the file
    file, err = os.Create(filename)
    if err != nil {
        fmt.Println("Error creating file:", err)
        return changeCounter
    }
    defer file.Close()
    writer := bufio.NewWriter(file)
    for _, line := range newLines {
        fmt.Fprintln(writer, line)
    }
    writer.Flush()
    return changeCounter
}

func main() {
    path := "./" // Current directory
    totalChangeCount := 0
    err := filepath.Walk(path, func(path string, info os.FileInfo, err error) error {
        if err != nil {
            fmt.Println("Error walking the path:", err)
            return err
        }
        if strings.HasSuffix(path, ".c") {
            fmt.Println("Processing file:", path)
            fileChangeCount := processFile(path)
            fmt.Println("Done! Changes done:", fileChangeCount)
            totalChangeCount += fileChangeCount
        }
        return nil
    })

    if err != nil {
        fmt.Println("Error walking the directory:", err)
    }

    fmt.Println("Done processing files! Total changes done:", totalChangeCount)
}
